Add get_site function and associated registry

eteq opened the pull request:
This is essentially a port of @bmorris3's work in astroplanners/astroplan#45 over to astropy.  The key element is that it adds a `get_site` function that returns an `EarthLocation` taken from a list of named observatories.

The actual data store is a JSON file populated from the IRAF obsdb file.  This is probably quite out of date, but is a good starting point as it's something many astronomers got use to having from the IRAF era.

I'd say this closes #3813, at least as an initial implementation - more is discussed there in terms of querying remote sources, but I think that is an "extra" layer that could be added on top of this.

cc @astrofrog @mhvk @bsipocz @cdeil as participants in #3813

astrofrog commented:
This will be very useful!

However, this adds and removes many lines/files over the commits unrelated to this PR - I wonder whether you could squash the commits a little while preserving a couple of commits for each author?

mhvk commented:
Sorry, did not have much of a look, but one simple addition: have a class method `EarthLocation.get_site` (as well).

embray commented:
Perhaps the observatories.json could also be added to data.astropy.org, along with mechanisms to automatically download and use updates when they become available (similar to what we do with IERS data)?

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+    if _site_db is None:
+        _load_sites()
+
+    if site_name.lower() not in _site_db.keys():
```
No need for `.keys()`

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+
+    if site_name.lower() not in _site_db.keys():
+        # If site name not found, find close matches and suggest them in error
+        close_names = get_close_matches(site_name, _site_db.keys())
```
Here too--any iterable can be passed in to `get_close_matches`.

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+        # If site name not found, find close matches and suggest them in error
+        close_names = get_close_matches(site_name, _site_db.keys())
+        close_names = sorted(close_names, key=lambda x: len(x))
+        if len(close_names) > 0:
```
Can just say `if close_names:`

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+        if len(close_names) > 0:
+            errmsg = ("Site not in database. Use ``get_site_names()`` "
+                      "to see available sites. Did you mean: '{}'?".format(
+                      "', '".join(close_names)))
```
Implicit format strings won't work on Python 2.6 (i.e. `{}` -> `{0}`).  Maybe also it would be clearer if it said "Did you mean one of: ...?"

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+                      "to see available sites. Did you mean: '{}'?".format(
+                      "', '".join(close_names)))
+        else:
+            errmsg = 'Site not in database.'
```
Maybe should include the requested site name in the error message.

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+    ----------
+    full_list : bool
+        Show full list observatory names and aliases (True), or just the list
+        of names (False)? Default to True.
```
I think should be named something more explicit like `show_aliases`.  `full_list` makes me think "Is it not showing me all the observatories by default for some reason?"

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+    remove_site : remove a site by name from the list of sites
+    """
+    if _site_db is None:
+        _load_sites()
```
I think a lot of boilerplate can be avoided if `_load_sites()` also actually returns the site DB.  So you can just do `sites = _load_sites()`

Alternatively this could be reorganized into a class that could in principle allow reuse for custom observatory DBs (for example maybe [PANOPTES](http://projectpanoptes.org/v1/)) would like its own observatory DB?  The module level functions could just be aliases to methods on a default observatory DB that exists at the module level (the class could easily handle all the lazy-loading).

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+
+def add_site(site_names, location):
+    """
+    Add a site to the list of available observatories.
```
Which makes me think, in addition to updating the DB from an online resource, this should include the ability to persist a custom DB to `~/.astropy` or elsewhere.  But something like that could be added later (though makes another good argument for reorganization into a class).

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+                firstnamedone = True
+        else:
+            raise KeyError('The site "{}" already exists at (longitude,latitude,'
+                           'elevation)={}'.format(name,
```
Cool--I didn't even know about this syntax (does it work on Python 2.x?)

eteq replied:
It does work on 2.7, but not 2.6, sadly. So I'll switch it to the standard "{0}" and "{1}"

embray commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,203 @@
+    site_name : string
+        Name of the observatory to remove
+    remove_aliases : bool
+        Also remove any aliases for the corresponding location
```
I think if you're going to remove a site it should just always remove all aliases.  A separate function(s) could be added to remove(/add) aliases.

eteq replied:
Yeah, I see your point here.

embray commented:
@mhvk 
> Sorry, did not have much of a look, but one simple addition: have a class method EarthLocation.get_site (as well).

Might be better if `EarthLocation.from_site` (though `EarthLocation.of_site()` has a certain ring to it too...).  If my suggestion for alternate DBs is implemented this should take an optional observatory DB argument as well.

eteq commented:
@embray @astrofrog @mhvk - OK, the version I just pushed up here integrates @embray's inline suggestions *and* makes a couple substansive changes:
* It removes add/remove capability.  After thinking about this and some of the discussion in #3813, it seems to me that the point here is to add a *convenience* method, akin to the `SkyCoord.from_name` functionality, rather than a fully-featured Site DB (which is one of the directions @embray suggested).  My thinking is that any application that would want to make their own Site DB would probably be better off just making their own `EarthLocation`s from scratch anyway.
* It removes the `sites` module from the public API in favor of only going through EarthLocation classmethods (which also makes it closer to `from_name`).
* It *adds* the capability to go and try to download the list of databases from ``http://data.astropy.org/locations.json``.  The `EarthLocation` version tries to do this but then falls back on the builtin version if it can't get online. Of course it will always do that right now, but we can immediately upload the current JSON file to data.astropy.org once this is merged (if we agree with this approach)

Sound good?

astrofrog commented:
@eteq - this looks good but is missing documentation. It deserves an entry in the main docs? (not just in API docs).

Also, there are some failures on Travis with Python 3.x. Are you still using Legacy Python?? :wink:

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -174,6 +179,73 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+        from .sites import get_site_names
+
+        try:
+            return get_site_names(show_aliases, online=True)  # this is always an EarthLocation
```
I don't understand the comment `# this is always an EarthLocation` given that this function returns a `list of str`.

eteq replied:
oops, good catch, that was a copy-and-paste error.

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,138 @@
+    Load observatory database from data/observatories.json and parse them into
+    a dictionary in memory.
+    """
+    if _builtin_site_dict.get('_site_db_uninitialized', False):
```
Wouldn't it be easier to just test `if not _buildin_site_names:`?
With that the whole `_site_db_uninitialized` key stuff could be removed.

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,138 @@
+"""
+Observatories accessible by the `sites` module originate from the IRAF
+Observatory Database, and are stored in ``data/observatories.json``.
+Longitudes are listed with positive to the West.
```
This confused the heck out of me, since I couldn't find where the sign is changed (since default is positive to east). But then I saw that the json actually has longitudes between 0 and 360, measured east of greenwich. Is this indeed the case?  If so, does this mean it changed the original IRAF file?  In any case, please correct this comment!

bmorris3 replied:
This is a comment that was written for `astroplan` which should probably be removed for `astropy`.

eteq replied:
Good catch - I'm going to be switching to the newer astroplan one in a commit I send up shortly, anyway.

mhvk commented:
Plans seems good.  PR looks good too, modulo the minor comments I just inserted. 

It does look like eventually we should have `EarthLocation` carry a name. In principle, since it already carries the `info = QuantityInfo()` attribute, this could be set in `info.name` (see in-line comment for where).

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -174,6 +179,73 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+            warn(AstropyUserWarning(_NO_ONLINE_SITES_WARNING_MSG))
+            el = get_site(site_name, online=False)  # this is always an EarthLocation
+
+        if cls is el.__class__:
```
If we want to carry the name, this could be replaced with:
```
if cls is not el.__class__:
    el = cls.from_geodetic(*el.to_geodetic())

el.info.name = site_name
return el
```

eteq replied:
I went with a slightly different implementation that actually uses the name from the db file (which is sometimes different), but basically the same idea.

embray commented on astropy/coordinates/earth.py:
```diff
@@ -20,6 +22,9 @@
 # Available ellipsoids (defined in erfam.h, with numbers exposed in erfa).
 ELLIPSOIDS = ('WGS84', 'GRS80', 'WGS72')
 
+_NO_ONLINE_SITES_WARNING_MSG = ('Could not access the online site list. Falling'
```
This is missing a space between "falling" and "back".

As a general policy when splitting a strong across lines I always add a space at the end of the line so that I don't forget.

eteq replied:
Good point, will fix
I *usually* do that too but in this case was trying to avoid the 80 chars (silly, really, but it's sort of a little game...)

embray commented on astropy/coordinates/earth.py:
```diff
@@ -174,6 +179,73 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+
+        try:
+            el = get_site(site_name, online=True)  # this is always an EarthLocation
+        except URLError:
```
I think there are lots of other classes of errors that could occur here.

eteq replied:
Hmm, I *thought* (but could be wrong) that `URLError` is the base error to try to catch here.  `urlopen` says it throws things that are  `URLError` or subclasses, anyway. I guess I could just catch *everything*... Do you think that's better, @embray?

embray replied:
I'm not entirely sure it should do this at all to be honest.  Silently updating from an online resource would be a surprising side-effect to me.  That's why I originally suggested following the pattern we do for IERS data, and allowing manual loading of specific DBs from a URL.

eteq commented:
OK, I think I've addressed all of the comments above, but inspired by @embray's comment, I did I refactoring to use a `SiteRegistry` class instead of the previous functions that depended on global state. 

None of that refactoring is in the public API right now, but the idea is that if we want to actually make this more visible/fully featured in astropy v1.2 we can do so then. I'm thinking the best way to move on that is to have astroplan be the guinea pig for that (cc @bmorris3), and then we see based on that whether to make it truly public.

bmorris3 commented:
@eteq – sounds good. Feel free to assign me to an issue to do this in astroplan when you're ready!

eteq commented:
Oh, and one other change in the most recent set of commits that may or may not be relevant - I updated the json file to be the newer astroplan file (as @bmorris3 suggested in #3813)

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,115 @@
+    @classmethod
+    def from_json(cls, jsondb):
+        reg = cls()
+        for site in jsondb:
```
Is `jsondb` sufficiently like a dict that one could iterate over both `site` and `jsondb[site]`? Otherwise, the latter is used so much it should be worth doing `site_info = jsondb[site]` and using that below.

eteq replied:
good point - will use the `site_info` approach.

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,115 @@
+
+        return self._site_dict[site_name.lower()]
+
+    def get_site_names(self):
```
Isn't this more logical as a property `site_names`?

mhvk replied:
Or just `names`...

eteq replied:
agreed - a `names` property makes the most sense here

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,115 @@
+    """
+    def __init__(self):
+        # the keys to this are always lower-case
+        self._site_dict = {}
```
I'm starting to nit-pick, but since this is already a site registry, I'd call these `_locations` and `_names` (certainly no need to append `_dict` -- it *is* a dict, after all!).

eteq replied:
I see your point here - I agree on `_names`, but after thinking a bit decided `_lowercase_names_to_locations` is a good name for the dict to make it very clear exactly what it is

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,115 @@
+        # these can be whatever case is appropriate
+        self._site_names = []
+
+    def get_site(self, site_name):
```
OK, realising that I now am becoming probably too irritating, but why not just define `__getitem__` instead? Indeed, one could even just subclass `dict`, i.e., `SiteRegistry(dict)` (or maybe `OrderedDict`).  Then, `add_site` would be replaced by an overload of `update`.

eteq replied:
I'd prefer to avoid actually overloading `update`, because of the possible subtlties in deciding how to deal with updates where there are case-inconsistent keys.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -174,6 +176,125 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+        -------
+        reg : astropy.coordinates.sites.SiteRegistry
+        """
+        # need to import inside function to avoid circular dependencies
```
I learned from @embray that one trick to avoid this is to put the import at the very end of the file. I *think* it should work here too.

eteq replied:
Ooh, I'll try that - didn't know that one!

mhvk commented:
I like the change to a site registry class but have some comments about the implementation (see inline). Mostly it seems that `SiteRegistry` is so close to a `dict` that one should just subclass it (and make the method names simpler).

eteq commented:
@mhvk - I agree with some of that, but I'm not so sure about making it into a dict.  Here's my reasoning: it has a subtle difference from dict in that the "keys" (the `_site_names`) have case, but the dict itself (`_site_dict`) are meant to be case-insensitive.  That is, ``EarthLocation.of_site('keck')`` should do the same as  ``EarthLocation.of_site('keCk')``.  This is mainly because this is supposed to be a "quick" function, so we don't want to be too picky about this stuff.  I thought that making it an actually mapping/dict-like interface would cause this subtle but important distinction to get lost.  It would also mean that there's not a one-to-one mapping from "keys" (the `_site_names`) to the actual dictionary keys.

If you think I'm worrying too much over that confusion feel free to say so, though - my opinion is not very strong here, especially given that it's currently all internal.

mhvk commented:
Hmm, yes, perhaps not completely dict-like, and if one would wind up just overwriting all the methods,  subclassing might not make that much sense any more. Though I think it still would make sense to use, as much as possible, the interface also presented by a `dict`, e.g., have a `__getitem__` and a `__contains__` (both of which apply `.lower()` on the input).

eteq commented:
@mhvk - ok, I think I've addressed your comments.  I changed it to a dict-like thing as you suggested *mostly*, but decided to maintain `add_site` because it seemed like that it would be much more confusing to deal with mixed-case situations via `update`.  So this should be good to go (if the tests pass...)

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,134 @@
+            close_names = get_close_matches(site_name, self._lowercase_names_to_locations)
+            close_names = sorted(close_names, key=lambda x: len(x))
+            if close_names:
+                errmsg = ('Site not in database. Use ``get_names()`` '
```
Sorry to come to this only now, and so trivial too, but shouldn't the error messages be identical except for the addition of suggested names? I.e., here, shouldn't you list what name was put in, and below, also give the hint of using `get_names()` to see the full list? If so, just make a standard message and append "Did you mean one of ..." for the case of close names.

eteq replied:
Good point, will do.

mhvk commented on astropy/coordinates/sites.py:
```diff
@@ -0,0 +1,134 @@
+        site : `~astropy.coordinates.EarthLocation`
+            The location of the observatory.
+        """
+        if site_name.lower() not in self._lowercase_names_to_locations:
```
This could now be `if site_name not in self`... (but definitely don't have too).

mhvk commented on docs/coordinates/index.rst:
```diff
@@ -199,12 +199,25 @@ for a particular named object::
 
-.. note::
+For sites (primarily observatories) on the Earth, `astropy.coordinates` provides
+a similar quick way to get an `~astropy.coordinates.EarthLocation` for a site
```
I'd delete "for a site given a name" (the sentence starts with "For sites...").

mhvk commented:
Looks good! Minor comment still about the error message (and two really tiny ones that you can do with what you want). Otherwise, good to go.

eteq commented:
Alright, integrated the last of @mhvk's changes.  I'll go ahead and merge this now then!

bmorris3 commented:
:clap: :+1:

bsipocz commented:
@eteq - :+1:  to move this into core.
However I'm sorry I didn't followed this one given that I had strong feelings about hard wiring the ancient IRAF obsdb. One thing got fixed compared to that already (like longitude is positive towards East unlike in the iraf file). 
The other issues I have with it is the minor inaccuracies and outdated status and it's quite US focused. E.g. there is an "eso" entry for La Silla, but neither Paranal nor Chajnantor is in there. Also the name both for mmt and flwo is "Whipple Observatory". Obviously they on the same mountain, but I think the "Name" field should be unique, too. 
Should I open a PR to fix at least the eso stuff?

eteq commented:
@bsipocz - my thinking here is that the built-in locations are really just examples - the primary "database" is the one that lives at https://github.com/astropy/astropy-data .  So I would advocate updating *that* instead of the built-in one.  Of course this is up for debate if you really don't like it, but I'd rather keep the bundled one intentionally very limited and have the online one (which will always be more up-to-date) be the "real" one.

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -174,6 +176,122 @@ def from_geodetic(cls, lon, lat, height=0., ellipsoid=None):
+        if cls is el.__class__:
+            return el
+        else:
+            neweel = cls.from_geodetic(*el.to_geodetic())
```
Ouch! While viewing `earth.py` in emacs, this line had some nice red squiggly lines under it -- clearly this should be `newel` -- this also means there is no test case for this line...  @eteq - care to make a quick PR to correct this, or raise a new issue? Sorry I missed this before...

embray replied:
Mmm eel...

eteq replied:
Not just any eel: *new* eel.  If only there were an emoji for that...

Will issue a PR shortly.

bmorris3 replied:
A PR for a typo correction or for a fresh eel emoji? On second though don't answer that, it might be disappointing.
